tests/vanilla/effect.test.ts

```diff
@@ -0,0 +1,230 @@
+import { expect, it, vi } from 'vitest'
+import type { Atom, Getter, Setter } from 'jotai/vanilla'
+import { atom, createStore } from 'jotai/vanilla'
+
+type AnyAtom = Atom<unknown>
+type GetterWithPeak = Getter & { peak: Getter }
+type SetterWithRecurse = Setter & { recurse: Setter }
+type Cleanup = () => void
+type Effect = (get: GetterWithPeak, set: SetterWithRecurse) => void | Cleanup
+type Ref = {
+  get: GetterWithPeak
+  set?: SetterWithRecurse
+  cleanup?: Cleanup | null
+  fromCleanup: boolean
+  inProgress: number
+  deps: Set<AnyAtom>
+  init: () => void
+}
+
+function atomSyncEffect(effect: Effect) {
+  const refAtom = atom(
+    () => ({ deps: new Set(), inProgress: 0 }) as Ref,
+    (get) => {
+      const ref = get(refAtom)
+      return () => {
+        ref.cleanup?.()
+        ref.cleanup = null
+        ref.deps.clear()
+      }
+    },
+  )
+  refAtom.onMount = (mount) => mount()
+  const internalAtom = atom((get) => {
+    const ref = get(refAtom)
+    if (!ref.get) {
+      ref.get = ((a) => {
+        ref.deps.add(a)
+        return get(a)
+      }) as Getter & { peak: Getter }
+    }
+    ref.init()
+    ref.deps.forEach(get)
```

Comment:
note: IIRC this hack was added in response to a breaking behavioral change this past fall dealing with promise values losing their dependencies between runs. But since the effect is now synchronous, we might be able to remove ref.deps.